wallet-new/src/Cardano/Wallet/API/Response.hs

```diff
@@ -91,16 +79,16 @@ instance (Arbitrary a, Arbitrary b) => Arbitrary (OneOf a b) where
 respondWith :: (Foldable f, Monad m)
             => RequestParams
             -> (RequestParams -> m (f a))
-            -> m (OneOf [a] (ExtendedResponse [a]))
+            -- ^ A callback-style function which, given the full set of `RequestParams`
+            -- produces some form of results in some 'Monad' @m@.
+            -> m (WalletResponse [a])
 respondWith params@RequestParams{..} generator = do
```

Comment:
Since we are already creating abstractions before they are actually used, maybe use something like:
```
class WalletResp a where
    size        :: Int
    totalPages  :: Int
```
Where we can abstract that over both a regular list `[a]` and something like a `Map k v` since we have a KV storage.
We can then have:
```
instance WalletResp [a] where
...
```
And something like:
```
instance WalletResp (Map k v) where
...
```
Bear in mind, these are initial ideas, can be improved.

Reply:
@ksaric I don't want to rub you the wrong way, but I have a bit of an aversion for typeclasses used to encapsulate some form of interface 😬 😬 . In general, I resort to them when I want to stick to SPJ's definition of a way to achieve "bounded polymorphism" or when the program structure dictates it (for example `Migrate` is an example where it might make sense to have it).

For things like `WalletResponse` I think a plain record is better: It makes trivial for a newcomer to understand what's going on rather than having all these different instances scattered throughout the program which tell how to build a response. In general, I think [explicit is better than implicit](https://www.python.org/dev/peps/pep-0020/#the-zen-of-python) and I'm willing to pay some price for it (in terms of boilerplate) 😉